external/avro/src/main/scala/org/apache/spark/sql/avro/AvroEncoder.scala

```diff
@@ -0,0 +1,549 @@
+    value = new Parser().parse(in.readObject().asInstanceOf[String])
+  }
+  override def writeExternal(out: ObjectOutput): Unit = out.writeObject(value.toString)
+  def resolveUnion(datum: Any): Int = GenericData.get.resolveUnion(value, datum)
+}
+
+object AvroExpressionEncoder {
+
+  def of[T <: SpecificRecord](avroClass: Class[T]): ExpressionEncoder[T] = {
+    val schema = avroClass.getMethod("getClassSchema").invoke(null).asInstanceOf[Schema]
+    assert(toSqlType(schema).dataType.isInstanceOf[StructType])
+    val serializer = AvroTypeInference.serializerFor(avroClass, schema)
+    val deserializer = AvroTypeInference.deserializerFor(schema)
+    new ExpressionEncoder[T](
+      serializer,
+      deserializer,
+      ClassTag[T](avroClass))
+  }
+
+  def of[T <: IndexedRecord](schema: Schema): ExpressionEncoder[T] = {
+    assert(toSqlType(schema).dataType.isInstanceOf[StructType])
+    val avroClass = Option(ReflectData.get.getClass(schema))
+      .map(_.asSubclass(classOf[SpecificRecord]))
+      .getOrElse(classOf[GenericData.Record])
+    val serializer = AvroTypeInference.serializerFor(avroClass, schema)
+    val deserializer = AvroTypeInference.deserializerFor(schema)
+    new ExpressionEncoder[T](
+      serializer,
+      deserializer,
+      ClassTag[T](avroClass))
+  }
+}
+
+/**
+ * Utilities for providing Avro object serializers and deserializers
+ */
+private object AvroTypeInference {
+
+  /**
+   * Translates an Avro Schema type to a proper SQL DataType. The Java Objects that back data in
+   * generated Generic and Specific records sometimes do not align with those suggested by Avro
+   * ReflectData, so we infer the proper SQL DataType to serialize and deserialize based on
+   * nullability and the wrapping Schema type.
+   */
+  private def inferExternalType(avroSchema: Schema): DataType = {
+    toSqlType(avroSchema) match {
+      // the non-nullable primitive types
+      case SchemaType(BooleanType, false) => BooleanType
+      case SchemaType(IntegerType, false) => IntegerType
+      case SchemaType(LongType, false) =>
+        if (avroSchema.getType == UNION) {
+          ObjectType(classOf[java.lang.Number])
+        } else {
+          LongType
+        }
+      case SchemaType(FloatType, false) => FloatType
+      case SchemaType(DoubleType, false) =>
+        if (avroSchema.getType == UNION) {
+          ObjectType(classOf[java.lang.Number])
+        } else {
+          DoubleType
+        }
+      // the nullable primitive types
+      case SchemaType(BooleanType, true) => ObjectType(classOf[java.lang.Boolean])
+      case SchemaType(IntegerType, true) => ObjectType(classOf[java.lang.Integer])
+      case SchemaType(LongType, true) => ObjectType(classOf[java.lang.Long])
+      case SchemaType(FloatType, true) => ObjectType(classOf[java.lang.Float])
+      case SchemaType(DoubleType, true) => ObjectType(classOf[java.lang.Double])
+      // the binary types
+      case SchemaType(BinaryType, _) =>
```

Comment:
When avroSchema is a Union{null, Fixed}, it results in a java.io.ByteBuffer. Have we considered this case?